### PR TITLE
hive: added hiveconfig.go to pkg 

### PR DESCRIPTION
**Pull request by ElaiShalevRH**

### CODE REVIEW

File: pkg/hive/hiveconfig.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+// HiveConfigAdditionalOptions additional options for HiveConfig object.
+type HiveConfigAdditionalOptions func(builder *HiveConfigBuilder) (*HiveConfigBuilder, error)
+
+// NewHiveConfigBuilder creates a new instance of HiveConfigBuilder.
+func NewHiveConfigBuilder(apiClient *clients.Settings, name string) *HiveConfigBuilder {
+	glog.V(100).Infof(
+		`Initializing new HiveConfig structure with the following params: name: %s, releaseImage: %s`,
+		name)
```
Comment (trewest) [inline pkg/hive/hiveconfig.go:32]:
```suggestion
		`Initializing new HiveConfig structure with the following params: name: %s`, name)
```

File: pkg/hive/hiveconfig.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// HiveConfigBuilder provides struct for the HiveConfig object containing connection to
+// the cluster and the HiveConfig definitions.
+type HiveConfigBuilder struct {
+	Definition *hiveV1.HiveConfig
+	Object     *hiveV1.HiveConfig
+	errorMsg   string
+	apiClient  *clients.Settings
```
Comment (trewest) [inline pkg/hive/hiveconfig.go:22]: can you add 
```
import (
   runtimeClient "sigs.k8s.io/controller-runtime/pkg/client"
)
```
And change this to
```suggestion
	apiClient  runtimeClient.Client
```

File: pkg/hive/hiveconfig.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+type HiveConfigAdditionalOptions func(builder *HiveConfigBuilder) (*HiveConfigBuilder, error)
+
+// NewHiveConfigBuilder creates a new instance of HiveConfigBuilder.
+func NewHiveConfigBuilder(apiClient *clients.Settings, name string) *HiveConfigBuilder {
+	glog.V(100).Infof(
+		`Initializing new HiveConfig structure with the following params: name: %s, releaseImage: %s`,
+		name)
+
+	builder := HiveConfigBuilder{
+		apiClient: apiClient,
```
Comment (trewest) [inline pkg/hive/hiveconfig.go:35]:
```suggestion
		apiClient: apiClient.Client,
```

File: pkg/hive/hiveconfig.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+	return builder
+}
+
+// PullHiveConfig loads an existing HiveConfig into HiveConfigBuilder struct.
+func PullHiveConfig(apiClient *clients.Settings, name string) (*HiveConfigBuilder, error) {
+	glog.V(100).Infof("Pulling existing HiveConfig name: %s", name)
+
+	builder := HiveConfigBuilder{
+		apiClient: apiClient,
```
Comment (trewest) [inline pkg/hive/hiveconfig.go:90]:
```suggestion
		apiClient: apiClient.Client,
```

**Review by trewest — COMMENTED** [review]
Just a few comments. These changes are part of an current initiative to limit the API clients passed into our structs

File: pkg/hive/hiveconfig.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+	apiClient  runtimeClient.Client
+}
+
+// HiveConfigAdditionalOptions additional options for HiveConfig object.
+type HiveConfigAdditionalOptions func(builder *HiveConfigBuilder) (*HiveConfigBuilder, error)
+
+// NewHiveConfigBuilder creates a new instance of HiveConfigBuilder.
+func NewHiveConfigBuilder(apiClient *clients.Settings, name string) *HiveConfigBuilder {
+	glog.V(100).Infof(
+		`Initializing new HiveConfig structure with the following params: name: %s, releaseImage: %s`, name)
```
Comment (trewest) [inline pkg/hive/hiveconfig.go:31]:
```suggestion
		`Initializing new HiveConfig structure with the following params: name: %s`, name)
```

**Review by trewest — CHANGES_REQUESTED** [review]
Looks good, just a couple of minor comments

**Comment by trewest** [comment]
cc @achuzhoy

**Comment by ElaiShalevRH** [comment]
cc @achuzhoy

File: pkg/hive/hiveconfig.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+	}, HiveConfig)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return HiveConfig, err
+}
+
+// Create generates a HiveConfig on the cluster.
```
Comment (achuzhoy) [inline pkg/hive/hiveconfig.go:124]: Note that hiveconfig gets autocreated.
If you delete it - it gets autocreated.
Do we need this method?

Comment (trewest) [reply pkg/hive/hiveconfig.go:124]: I was also wondering about that but couldn't find anything that stood out in their repo. I think we can remove it as I have never had to manually create one

Comment (ElaiShalevRH) [reply pkg/hive/hiveconfig.go:124]: @trewest testing

Comment (trewest) [reply pkg/hive/hiveconfig.go:124]: @ElaiShalevRH You should be safe to remove the Create method. Just change your update method to align with this one: https://github.com/openshift-kni/eco-goinfra/blob/main/pkg/deployment/deployment.go#L395

**Comment by ElaiShalevRH** [comment]
@trewest @achuzhoy 
last changes

**Review by trewest — APPROVED** [review]
/lgtm

**Comment by achuzhoy** [comment]
@trewest   do we not want to limit the apiclient for pull/new in this package?

File: pkg/hive/hiveconfig.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+// HiveConfigAdditionalOptions additional options for HiveConfig object.
+type HiveConfigAdditionalOptions func(builder *HiveConfigBuilder) (*HiveConfigBuilder, error)
+
+// NewHiveConfigBuilder creates a new instance of HiveConfigBuilder.
+func NewHiveConfigBuilder(apiClient *clients.Settings, name string) *HiveConfigBuilder {
+	glog.V(100).Infof(
+		`Initializing new HiveConfig structure with the following params: name: %s`, name)
+
```
Comment (trewest) [inline pkg/hive/hiveconfig.go:32]:
```suggestion
	if apiClient == nil {
		glog.V(100).Infof("The apiClient cannot be nil")

		return nil
	}
```

File: pkg/hive/hiveconfig.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+		}
+	}
+
+	return builder
+}
+
+// PullHiveConfig loads an existing HiveConfig into HiveConfigBuilder struct.
+func PullHiveConfig(apiClient *clients.Settings, name string) (*HiveConfigBuilder, error) {
+	glog.V(100).Infof("Pulling existing HiveConfig name: %s", name)
+
```
Comment (trewest) [inline pkg/hive/hiveconfig.go:81]:
```suggestion
	if apiClient == nil {
		glog.V(100).Infof("The apiClient cannot be nil")

		return nil
	}
```

**Comment by trewest** [comment]
> @trewest do we not want to limit the apiclient for pull/new in this package?

@achuzhoy It's  already limited. Just added suggestions for checking if apiClient is nil though

**Review by achuzhoy — APPROVED** [review]

**Pull request merged**